handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -917,16 +927,19 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
         discardSomeReadBytes();
 
         flushIfNeeded(ctx);
+        readIfNeeded(ctx);
+
+        firedChannelRead = false;
+        ctx.fireChannelReadComplete();
+    }
 
+    private void readIfNeeded(ChannelHandlerContext ctx) {
         // If handshake is not finished yet, we need more data.
         if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
             // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
             // yet, which means we need to trigger the read to ensure we not encounter any stalls.
             ctx.read();
```

Comment:
> I'd probably need to add throws Exception for every single method in the call-graph

@vkostyukov - You should be able to remove `throws Exception` from the signature of `public void read(ChannelHandlerContext ctx)` ... however to your point it seems reasonable that a read "synthesized" by SslHandler should not force a read after the handshake is complete.